### ISSUE TITLE
Segmentation fault

### ISSUE THREAD

**Issue by a046829713** [state-spaces/mamba#725]
Hello,

I’m very happy to be using your library, but I keep encountering the following error during computation. Could you please offer some guidance? I’ve been testing and researching for months and really can’t find a solution, so I’m reaching out now.

I am using mamba version one model.

GPU Version
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.144                Driver Version: 570.144        CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4080 ...    Off |   00000000:01:00.0 Off |                  N/A |
|  0%   38C    P8              5W /  320W |     167MiB /  16376MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            1497      G   /usr/lib/xorg/Xorg                      122MiB |
|    0   N/A  N/A            1709      G   /usr/bin/gnome-shell                      8MiB |
+-----------------------------------------------------------------------------------------+

nvcc version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2025 NVIDIA Corporation
Built on Fri_Feb_21_20:23:50_PST_2025
Cuda compilation tools, release 12.8, V12.8.93
Build cuda_12.8.r12.8/compiler.35583870_0

package List
causal-conv1d            1.5.0.post8
certifi                  2025.4.26
charset-normalizer       3.4.1
cloudpickle              3.1.1
einops                   0.8.1
Farama-Notifications     0.0.4
filelock                 3.13.1
fsspec                   2024.6.1
gymnasium                0.29.1
huggingface-hub          0.30.2
idna                     3.10
Jinja2                   3.1.4
mamba_ssm                2.2.4
MarkupSafe               2.1.5
mpmath                   1.3.0
networkx                 3.3
ninja                    1.11.1.4
numpy                    1.26.3
nvidia-cublas-cu12       12.8.3.14
nvidia-cuda-cupti-cu12   12.8.57
nvidia-cuda-nvrtc-cu12   12.8.61
nvidia-cuda-runtime-cu12 12.8.57
nvidia-cudnn-cu12        9.7.1.26
nvidia-cufft-cu12        11.3.3.41
nvidia-cufile-cu12       1.13.0.11
nvidia-curand-cu12       10.3.9.55
nvidia-cusolver-cu12     11.7.2.55
nvidia-cusparse-cu12     12.5.7.53
nvidia-cusparselt-cu12   0.6.3
nvidia-nccl-cu12         2.26.2
nvidia-nvjitlink-cu12    12.8.61
nvidia-nvtx-cu12         12.8.55
packaging                25.0
pandas                   2.2.3
pillow                   11.0.0
pip                      22.0.2
protobuf                 6.30.2
python-dateutil          2.9.0.post0
pytz                     2025.2
PyYAML                   6.0.2
regex                    2024.11.6
requests                 2.32.3
safetensors              0.5.3
setuptools               80.1.0
six                      1.17.0
sympy                    1.13.3
tensorboardX             2.6.2.2
tokenizers               0.21.1
torch                    2.7.0+cu128
torchaudio               2.7.0+cu128
torchvision              0.22.0+cu128
tqdm                     4.67.1
transformers             4.51.3
triton                   3.3.0
typing_extensions        4.12.2
tzdata                   2025.2
urllib3                  2.4.0
wheel                    0.45.1

Error List:
Fatal Python error: Segmentation fault

Current thread 0x00007dc38fa7c000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/dain.py", line 89 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/lib/model.py", line 460 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/agent.py", line 95 in __call__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 109 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 210 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/experience.py", line 72 in populate
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 296 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

Fatal Python error: Segmentation fault

Thread 0x0000791276f5d640 (most recent call first):
  <no Python frame>

Current thread 0x000079146c380000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/functional.py", line 1425 in dropout
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/dropout.py", line 70 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/ssm_tool.py", line 60 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/modules/block.py", line 86 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/ssm_tool.py", line 295 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/lib/model.py", line 462 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/agent.py", line 95 in __call__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 109 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 210 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/experience.py", line 72 in populate
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 296 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

Fatal Python error: Segmentation fault

Current thread 0x00007d8ca2ba0000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/lib/model.py", line 469 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/agent.py", line 95 in __call__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 109 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 210 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/experience.py", line 72 in populate
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 296 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

Fatal Python error: Segmentation fault

Current thread 0x0000767ce6e6c640 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/einops/_backends.py", line 93 in reshape
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/einops/einops.py", line 251 in _apply_recipe
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/einops/einops.py", line 532 in reduce
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/einops/einops.py", line 600 in rearrange
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/ops/selective_scan_interface.py", line 334 in backward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/amp/autocast_mode.py", line 556 in decorate_bwd
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/autograd/function.py", line 307 in apply

Thread 0x0000767ed57d8000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/autograd/graph.py", line 824 in _engine_run_backward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/autograd/__init__.py", line 353 in backward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/_tensor.py", line 648 in backward
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 328 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

Fatal Python error: Segmentation fault

Thread 0x00007d26baf5d640 (most recent call first):
  <no Python frame>

Current thread 0x00007d28af644000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/autograd/function.py", line 575 in apply
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/ops/triton/layer_norm.py", line 902 in layer_norm_fn
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/modules/block.py", line 57 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/ssm_tool.py", line 295 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/lib/model.py", line 462 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/agent.py", line 95 in __call__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 109 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 210 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/experience.py", line 72 in populate
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 296 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

Fatal Python error: Segmentation fault

Thread 0x0000792e12390640 (most recent call first):
  <no Python frame>

Current thread 0x0000793000113000 (most recent call first):
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/ops/selective_scan_interface.py", line 215 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/amp/autocast_mode.py", line 510 in decorate_fwd
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/autograd/function.py", line 575 in apply
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/ops/selective_scan_interface.py", line 377 in mamba_inner_fn
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/modules/mamba_simple.py", line 146 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/mamba_ssm/modules/block.py", line 67 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/ssm_tool.py", line 295 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/lib/model.py", line 462 in forward
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1762 in _call_impl
  File "/home/b0457812963/testmambapy2.6/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1751 in _wrapped_call_impl
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/agent.py", line 95 in __call__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 109 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/DQN/ptan/experience.py", line 210 in __iter__
  File "/home/b0457812963/testmambapy2.6/SynapseX/Brain/Common/experience.py", line 72 in populate
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 296 in train
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 263 in __init__
  File "/home/b0457812963/testmambapy2.6/SynapseX/DQN_rl_train.py", line 399 in <module>

**Comment by beotborry** [comment]
I got an relevant error, segmentation fault raised by workers.

I resolved the issue by adding the following code at the beginning of the main function.

```
    import torch.multiprocessing as mp
    mp.set_start_method('spawn', force=True)
```
